synapseclient/producer_consumer_download_threads.py

```diff
@@ -0,0 +1,282 @@
+from queue import Queue
+from threading import Thread
+import typing
+import time
+from typing import Generator
+from re import split
+from datetime import datetime
+from math import ceil
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+from synapseclient.core.utils import printTransferProgress
+import json
+import os
+import requests
+
+# constants
+BUF_SIZE = 20
+MAX_RETRIES = 5
+MB = 2**20
+SYNAPSE_DEFAULT_DOWNLOAD_PART_SIZE = 8 * MB
+SYNAPSE_DEFAULT_FILE_ENDPOINT = "https://repo-prod.prod.sagebase.org/file/v1"
+SYNAPSE_URL_PATH_DOWNLOAD_GET_BATCH_PRE_SIGNED_URL = '/fileHandle/batch'
+ISO_AWS_STR_FORMAT = '%Y%m%dT%H%M%SZ'
+PARTIAL_CONTENT_CODE = 206
+CONNECT_FACTOR = 3
+BACK_OFF_FACTOR = 0.5
+
+# transfer progress parameters
+transferred = 0
```

Comment:
Explore using "multiprocessing.value" to pass these values to threads and allow for safe mutation. Check if analogous resource for threads only with less overhead.

Reply:
Refactored, still using global variable, incorporated a lock.